torchtune/models/phi4/_tokenizer.py

```diff
@@ -0,0 +1,308 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Any, Dict, List, Mapping, Optional, Tuple
+
+from torchtune.data._messages import Message
+from torchtune.data._prompt_templates import PromptTemplate
+from torchtune.data._utils import truncate
+from torchtune.modules.tokenizers import GPT2BaseTokenizer, TikTokenBaseTokenizer, ModelTokenizer
+from torchtune.modules.transforms import Transform
+
+PHI4_SPECIAL_TOKENS = {
+    "<|dummy_0|>": 100256,
+    "<|endoftext|>": 100257,
+    "<|fim_prefix|>": 100258,
+    "<|fim_middle|>": 100259,
+    "<|fim_suffix|>": 100260,
+    "<|dummy_1|>": 100261,
+    "<|dummy_2|>": 100262,
+    "<|dummy_3|>": 100263,
+    "<|im_start|>": 100264,
+    "<|im_end|>": 100265,
+    "<|im_sep|>": 100266,
+    "<|endofprompt|>": 100276,
+}
+
+# Adding other <|dummy_x|>
+
+for token_id in range(100266, 100351):
+    if token_id == 100276:
+        continue
+    PHI4_SPECIAL_TOKENS[f"<|dummy_{87 - (100350 - token_id)}|>"] = token_id + 1
+
+CL100K_PATTERN = r"""(?i:'s|'t|'re|'ve|'m|'ll|'d)|[^\r\n\p{L}\p{N}]?\p{L}+|\p{N}{1,3}| ?[^\s\p{L}\p{N}]+[\r\n]*|\s*[\r\n]+|\s+(?!\S)|\s+"""  # noqa
+
+
+class Phi4MiniTokenizer(ModelTokenizer, Transform):
+    """
+    TikToken tokenizer configured with Phi4 (14B) special tokens.
+
+    Args:
+        merges_path (str): Path to merges.txt file.
```

Comment:
Need to add path field if we're supporting it (but also see my comment below, I'm unclear on why we need it)